packages/react/src/side-nav/SideNavLinkList.tsx

```diff
@@ -1,51 +1,121 @@
-import { SideNavLink } from './SideNavLink';
+import { useCallback, useMemo } from 'react';
+import { SideNavUnorderedList } from './SideNavUnorderedList';
+import { SideNavListItem } from './SideNavListItem';
 import { SideNavProps } from './SideNav';
 import { hasNestedActiveItem } from './utils';
+import { SideNavLink } from './SideNavLink';
 
 export type SideNavLinkListProps = {
 	activePath: string | undefined;
 	items: SideNavProps['items'];
+	showSubLevel: SideNavProps['showSubLevel'];
 };
 
-export function SideNavLinkList({
-	activePath,
-	items: itemsProp,
-}: SideNavLinkListProps) {
-	const items = useMemo(() => {
-		return itemsProp.map((item) => ({
+type ItemWithIsActive = Omit<SideNavLinkListProps['items'][number], 'items'> & {
+	isActive?: boolean;
+	items?: ItemWithIsActive[];
+};
+
+// Recursively add `isActive` to any sub-level items
+const addIsActive =
+	(activePath: SideNavLinkListProps['activePath']) =>
+	(item: ItemWithIsActive): ItemWithIsActive => {
+		const withNestedActiveItems = hasNestedActiveItem(item.items, activePath);
+		const isCurrentPage = item.href === activePath;
+		const isActive = isCurrentPage || withNestedActiveItems;
+
+		return {
 			...item,
-			isActive:
-				item.href === activePath || hasNestedActiveItem(item.items, activePath),
-		}));
-	}, [activePath, itemsProp]);
+			isActive,
+			items:
+				item.items?.length || isActive
+					? item?.items?.map(addIsActive(activePath))
+					: item.items,
+		};
+	};
+
+export const SideNavLinkList = ({
+	activePath,
+	items,
+	showSubLevel,
+}: SideNavLinkListProps) => {
+	const itemsWithIsActive = useMemo(() => {
+		return items.map(addIsActive(activePath));
+	}, [activePath, items]);
+
+	const isTopLevelItemActive =
+		itemsWithIsActive.find((item) => item.isActive)?.href === activePath;
 
-	return <LinkList activePath={activePath} items={items} />;
-}
+	return (
+		<LinkList
+			activePath={activePath}
+			isListOpen={isTopLevelItemActive}
+			items={itemsWithIsActive}
+			showSubLevel={showSubLevel}
+		/>
+	);
+};
 
 type LinkListProps = {
 	activePath: string | undefined;
-	items: (SideNavProps['items'][number] & { isActive?: boolean })[];
+	isListOpen: boolean;
+	items: ItemWithIsActive[];
+	showSubLevel: SideNavProps['showSubLevel'];
 };
 
-function LinkList({ activePath, items }: LinkListProps) {
+const LinkList = ({
+	activePath,
+	isListOpen,
+	items,
+	showSubLevel,
+}: LinkListProps) => {
+	const isOpen = useCallback(
+		(
+			items?: SideNavLinkListProps['items'] | ItemWithIsActive[],
+			isActive?: boolean
+		) =>
+			isActive ||
+			showSubLevel === 'always' ||
+			hasNestedActiveItem(items, activePath),
+		[activePath, showSubLevel]
+	);
+
 	return (
-		<SideNavGroup>
-			{items.map(({ isActive, items, ...item }, idx) => {
+		<SideNavUnorderedList isOpen={isOpen(items, isListOpen)}>
+			{items.map(({ isActive, items, ...item }) => {
+				const hasNestedItemsIndicator =
+					items && 'length' in items && showSubLevel === 'whenActive';
```

Comment:
```suggestion
				const hasNestedItemsIndicator =
					showSubLevel === 'whenActive' && items && 'length' in items;
```

JavaScript evaluates from left to right, so we can short-circuit this for `showSubLevel="always"` and stop immediately. Otherwise, that use-case always has to see if it has `items` when it's irrelevant